capsules/src/mac.rs

```diff
@@ -0,0 +1,365 @@
+//!
+//! TODO: Encryption/decryption
+//! TODO: Sending beacon frames
+//! TODO: Channel scanning
+
+use core::cell::Cell;
+use kernel::ReturnCode;
+use kernel::common::take_cell::TakeCell;
+use kernel::hil::radio;
+use net::ieee802154::*;
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub struct FrameInfo {
+    // These offsets are relative to buf[radio::PSDU_OFFSET..] so that
+    // the mac frame length is data_offset + data_len
+    mac_payload_offset: usize,
+    data_offset: usize,
+    data_len: usize,
+    security_params: Option<(Security, [u8; 16])>,
+}
+
+impl FrameInfo {
+    pub fn frame_length(&self) -> usize {
+        self.data_offset + self.data_len
+    }
+
+    pub fn remaining_data_capacity(&self, buf: &[u8]) -> usize {
+        buf.len() - radio::PSDU_OFFSET - radio::MFR_SIZE - self.data_offset - self.data_len
+    }
+
+    pub fn append_payload(&mut self, buf: &mut [u8], payload: &[u8]) -> ReturnCode {
+        if payload.len() > self.remaining_data_capacity(buf.as_ref()) {
+            return ReturnCode::ENOMEM;
+        }
+        let begin = radio::PSDU_OFFSET + self.data_offset + self.data_len;
+        buf[begin..begin + payload.len()].copy_from_slice(payload);
+        self.data_len += payload.len();
+
+        ReturnCode::SUCCESS
+    }
+}
+
+pub trait Mac {
+    fn get_address(&self) -> u16; //....... The local 16-bit address
+    fn get_address_long(&self) -> [u8; 8]; // 64-bit address
+    fn get_pan(&self) -> u16; //........... The 16-bit PAN ID
+    fn get_channel(&self) -> u8;
+    fn get_tx_power(&self) -> i8;
+
+    fn set_address(&self, addr: u16);
+    fn set_address_long(&self, addr: [u8; 8]);
+    fn set_pan(&self, id: u16);
+    fn set_channel(&self, chan: u8) -> ReturnCode;
+    fn set_tx_power(&self, power: i8) -> ReturnCode;
+
+    fn config_commit(&self) -> ReturnCode;
+
+    fn is_on(&self) -> bool;
+    fn prepare_data_frame(&self,
+                          buf: &mut [u8],
+                          dst_pan: PanID,
+                          dst_addr: MacAddress,
+                          src_pan: PanID,
+                          src_addr: MacAddress,
+                          security_needed: Option<(SecurityLevel, KeyId)>)
+                          -> Result<FrameInfo, ()>;
+    fn transmit(&self,
+                buf: &'static mut [u8],
+                frame_info: FrameInfo)
+                -> (ReturnCode, Option<&'static mut [u8]>);
+}
+
+pub trait TxClient {
+    fn send_done(&self, spi_buf: &'static mut [u8], acked: bool, result: ReturnCode);
+}
+
+pub trait RxClient {
+    fn receive<'a>(&self,
+                   buf: &'a [u8],
+                   header: Header<'a>,
+                   data_offset: usize,
+                   data_len: usize,
+                   result: ReturnCode);
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+enum TxState {
+    Idle,
+    Encrypting,
+    ReadyToTransmit,
+}
+
+pub struct MacDevice<'a, R: radio::Radio + 'a> {
```

Comment:
Document. I suspect this will mostly overlap with the module documentation, so a reference from here or to here is probably fine for the details.